### ISSUE TITLE
Can't catch "This database has already been closed" exception triggered by a transaction

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#1442]
Here's a small repro - our moor version is 3.31:

```dart
import 'dart:async';

import 'package:test/test.dart';
import 'package:moor/ffi.dart';
import 'package:moor/moor.dart';

class TestDb extends GeneratedDatabase {
  TestDb() : super(SqlTypeSystem.withDefaults(), VmDatabase.memory());
  @override final List<TableInfo> allTables = const [];
  @override final int schemaVersion = 1;
}

Future<int> selectOne(TestDb _db) => _db
  .customSelect('select 1 a')
  .map((row) => row.read<int>('a'))
  .getSingle();


void main() {
  for (final useTransaction in [false, true]) {
    test('query after db.close, useTransaction=$useTransaction', () async {
      final db = TestDb();
      expect(await selectOne(db), 1);
      await db.close();
      try {
        if (useTransaction) {
          await db.transaction(() => selectOne(db));
        }
        else {
          await selectOne(db);
        }
        fail('should throw because db is closed');
      } catch (e) {
        expect(e.toString(), contains('This database has already been closed'));
      }
    });
  }
}
```

The first iteration where `useTransaction==false` passes, but the second iteration throws with the following
<details><summary>stacktrace</summary>

```
package:sqlite3/src/impl/database.dart 290:7                                       DatabaseImpl._ensureOpen
package:sqlite3/src/impl/database.dart 85:5                                        DatabaseImpl.execute
package:moor/src/ffi/vm_database.dart 100:11                                       _VmDelegate._runWithArgs
package:moor/src/ffi/vm_database.dart 110:11                                       _VmDelegate.runCustom
package:moor/src/runtime/executor/helpers/engines.dart 79:19                       _ExecutorWithQueryDelegate.runCustom.<fn>
package:synchronized/src/basic_lock.dart 32:26                                     BasicLock.synchronized
package:moor/src/runtime/executor/helpers/engines.dart 22:26                       _ExecutorWithQueryDelegate._synchronized
package:moor/src/runtime/executor/helpers/engines.dart 76:12                       _ExecutorWithQueryDelegate.runCustom
package:moor/src/runtime/executor/helpers/engines.dart 145:15                      _TransactionExecutor.ensureOpen.<fn>
package:synchronized/src/basic_lock.dart 32:26                                     BasicLock.synchronized
package:moor/src/runtime/executor/helpers/engines.dart 22:26                       _ExecutorWithQueryDelegate._synchronized
package:moor/src/runtime/executor/helpers/engines.dart 143:21                      _TransactionExecutor.ensureOpen
package:moor/src/runtime/api/query_engine.dart 98:21                               QueryEngine.doWhenOpened
package:moor/src/runtime/query_builder/statements/select/custom_select.dart 53:19  CustomSelectStatement._executeWithMappedArgs
package:moor/src/runtime/query_builder/statements/select/custom_select.dart 37:12  CustomSelectStatement.get
package:moor/src/runtime/query_builder/statements/query.dart 143:20                _MappedSelectable.get
package:moor/src/runtime/query_builder/statements/query.dart 100:24                Selectable.getSingle
test\moor_test.dart 16:4                                                           selectOne
test\moor_test.dart 27:38                                                          main.<fn>.<fn>
package:moor/src/runtime/api/query_engine.dart 342:38                              QueryEngine.transaction.<fn>.<fn>
dart:async                                                                         runZoned
package:moor/src/runtime/api/query_engine.dart 406:12                              QueryEngine._runEngineZoned
package:moor/src/runtime/api/query_engine.dart 339:14                              QueryEngine.transaction.<fn>
package:moor/src/runtime/api/query_engine.dart 98:64                               QueryEngine.doWhenOpened.<fn>
dart:async                                                                         _completeOnAsyncReturn
package:synchronized/src/basic_lock.dart                                           BasicLock.synchronized
===== asynchronous gap ===========================
dart:async                                                                         _asyncErrorWrapperHelper
package:moor/src/runtime/executor/helpers/engines.dart 143:21                      _TransactionExecutor.ensureOpen
package:moor/src/runtime/api/query_engine.dart 98:21                               QueryEngine.doWhenOpened
package:moor/src/runtime/query_builder/statements/select/custom_select.dart 53:19  CustomSelectStatement._executeWithMappedArgs
package:moor/src/runtime/query_builder/statements/select/custom_select.dart 37:12  CustomSelectStatement.get
package:moor/src/runtime/query_builder/statements/query.dart 143:20                _MappedSelectable.get
package:moor/src/runtime/query_builder/statements/query.dart 100:24                Selectable.getSingle
test\moor_test.dart 16:4                                                           selectOne
test\moor_test.dart 27:38                                                          main.<fn>.<fn>
package:moor/src/runtime/api/query_engine.dart 342:38                              QueryEngine.transaction.<fn>.<fn>
dart:async                                                                         runZoned
package:moor/src/runtime/api/query_engine.dart 406:12                              QueryEngine._runEngineZoned
package:moor/src/runtime/api/query_engine.dart 339:14                              QueryEngine.transaction.<fn>
package:moor/src/runtime/api/query_engine.dart 98:64                               QueryEngine.doWhenOpened.<fn>
dart:async                                                                         _completeOnAsyncReturn
package:synchronized/src/basic_lock.dart                                           BasicLock.synchronized
===== asynchronous gap ===========================
dart:async                                                                         Future.then
package:moor/src/runtime/api/query_engine.dart 98:50                               QueryEngine.doWhenOpened
package:moor/src/runtime/api/query_engine.dart 335:27                              QueryEngine.transaction
test\moor_test.dart 27:20                                                          main.<fn>
dart:async                                                                         _completeOnAsyncReturn
package:moor/src/runtime/api/db_base.dart                                          GeneratedDatabase.close
===== asynchronous gap ===========================
dart:async                                                                         _asyncThenWrapperHelper
test\moor_test.dart                                                                main.<fn>

Bad state: This database has already been closed
```
</details>

**Comment by simolus3** [comment]
Thanks for opening this issue. It looks like this is already fixed  on `develop` (and I most likey by `4.5.0` on pub too), I think 4e6819aab90d649e934de484b79f2903c44a5407 is the commit that fixed this. I can take a look at backporting that if its helpful for you.

**Comment by Mike278** [comment]
Ah nice 👍 Don't worry about backporting, we only encountered it in tests and used another approach to workaround it. Thanks again!
